code/modules/mob/living/carbon/xenomorph/Powers.dm

```diff
@@ -1020,10 +1024,70 @@
 		amount = plasma_stored //Just use all of it
 	use_plasma(amount)
 	target.gain_plasma(amount)
-	to_chat(target, "<span class='xenowarning'>\The [src] has transfered [amount] plasma to you. You now have [target.plasma_stored].</span>")
-	to_chat(src, "<span class='xenowarning'>You have transferred [amount] plasma to \the [target]. You now have [plasma_stored].</span>")
+	to_chat(target, "<span class='xenowarning'>[src] has transfered [amount] units of [energy] to you. You now have [target.plasma_stored]/[target.xeno_caste.plasma_max].</span>")
+	to_chat(src, "<span class='xenowarning'>You have transferred [amount] units of [energy] to [target]. You now have [plasma_stored]/[xeno_caste.plasma_max].</span>")
 	playsound(src, "alien_drool", 25)
 
+/mob/living/carbon/Xenomorph/proc/xeno_salvage_plasma(atom/A, amount, salvage_delay, max_range)
+	if(!isXeno(A) || !check_state())
+		return
+
+	var/mob/living/carbon/Xenomorph/target = A
+	var/is_silicon = xeno_caste & CASTE_IS_ROBOTIC ? TRUE : FALSE
+	var/energy = is_silicon ? "power" : "plasma"
+
+	if(!isturf(loc))
+		to_chat(src, "<span class='warning'>You can't salvage [energy] from here!</span>")
+		return
+
+	if(is_silicon ? !(target.xeno_caste & CASTE_IS_ROBOTIC) : target.xeno_caste & CASTE_IS_ROBOTIC)
+		to_chat(src, "<span class='warning'>[target]'s energy sources are incompatible with ours.</span>")
+		return
+
+	if(target.stat != DEAD)
+		to_chat(src, "<span class='warning'>You can't steal [energy] from living sisters, ask for some to a drone or a hivelord instead!</span>")
+		return
+
+	if(get_dist(src, target) > max_range)
+		to_chat(src, "<span class='warning'>You need to be closer to [target].</span>")
+		return
+
+	if(!(target.plasma_stored))
+		to_chat(src, "<span class='notice'>[target] doesn't have any [energy] left to salvage.</span>")
+		return
+
+	to_chat(src, "<span class='notice'>You start salvaging [energy] from [target].</span>")
+
+	while(target.plasma_stored)
```

Comment:
Are you sure a while loop is the best way to do this? When those loops go wrong, they usually go /very/ wrong. Wouldn't it better so xenos have to salvage multiple times so no looping would be neccessary?